### PR TITLE
[FR] Add support to decouple actions and exceptions

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
part of elastic/ia-trade-team/issues/258

## Summary
Details forthcoming

### TODO

 - [ ] helper methods from within TomlRuleContents to help associate rule <=> exceptions
    - [ ] this includes the "build time" functionality from # 2 above

### CODE REVIEW

File: detection_rules/action.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    updated_date: definitions.Date
+
+    # Optional fields
+    deprecation_date: Optional[definitions.Date]
+    comments: Optional[str]
+    maturity: Optional[definitions.Maturity]
+
+
+@dataclass
+class Action(MarshmallowDataclassMixin):
```
Comment (brokensound77) [inline detection_rules/action.py:30]: TODo: update with https://www.elastic.co/guide/en/security/current/bulk-actions-rules-api.html#actions-object-schema-bulk

File: detection_rules/etc/_config.yaml [hunk 2]
```diff
@@ -18,6 +21,12 @@ files:
 #         ├── packages.yml
 #         ├── stack-schema-map.yaml
 #         └── version.lock.json
+#     └── actions
+##         ├── action_1.toml
+##         ├── action_2.toml
+#     └── exceptions
+##         ├── exception_1.toml
+##         ├── exception_2.toml
```
Comment (Mikaayenson) [inline detection_rules/etc/_config.yaml:29]: Does this mean users will have to manually list every actions/exceptions list file in the config ?

File: detection_rules/action.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+from .mixins import MarshmallowDataclassMixin
+from .schemas import definitions
+
+
+@dataclass(frozen=True)
+class ActionMeta(MarshmallowDataclassMixin):
+    """Data stored in an exception's [metadata] section of TOML."""
+    creation_date: definitions.Date
+    rule_id: definitions.UUIDString
```
Comment (Mikaayenson) [inline detection_rules/action.py:19]: Should action list be associated to only one rule or multiple rules?

**Comment by Mikaayenson** [comment]
@brokensound77 here is an example toml file based on the actions:

```
[metadata]
creation_date = "2024-02-21"
rule_id = "5d1e96c6-1ee8-4f19-9416-1d8d81428f59"
rule_name = "Example Rule Name"
updated_date = "2024-02-22"
deprecation_date = "2025-01-01"      # optional
comments = "This is an example action list" # optional
maturity = "beta"                    # optional

[[actions]]
action_type_id = ".email"
group = "default"
params.message = "Action triggered: Example Rule Name"
id = "action_001"                    # optional
frequency = { "throttle": "5m" }     # optional  

[[actions]]
action_type_id = ".slack"
group = "default"
params.message = "Some other notification"
```

I noticed that ActionParams may need to be updated and `alerts_filter` added.

**Comment by Mikaayenson** [comment]
Sample exception list file:

```
[metadata]
creation_date = "2024-02-21"
rule_id = "5d1e96c6-1ee8-4f19-9416-1d8d81428f59"
rule_name = "Example Rule Name"
updated_date = "2024-02-22"
comments = "This is an example exception list."
maturity = "development"

[[exceptions]]
description = "Example exception container"
list_id = "exception_list_01"
name = "Sample Exception List"
namespace_type = "single"
tags = ["tag1", "tag2"]
type = "detection"

  [[exceptions.items]]
  description = "Exception item description"
  list_id = "item_list_01"
  name = "Exception Item Name"
  namespace_type = "single"
  tags = ["exception_item_tag1"]
  type = "simple"
  expire_time = "2024-12-31T23:59:59Z"

    [[exceptions.items.entries]]
    field = "process.name"
    type = "match_any"
    operator = "included"
    value = ["malicious_process", "another_process"]

    [[exceptions.items.entries]]
    field = "nested.field.example"
    type = "nested"
    operator = "included"

      [[exceptions.items.entries.entries]]
      field = "nested.inside"
      type = "match"
      operator = "included"
      value = "specific_value"
```

File: detection_rules/generic_loader.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                         toml_filter: Optional[Callable[[dict], bool]] = None):
+        for path in directories:
+            self.load_directory(path, recursive=recursive, toml_filter=toml_filter)
+
+    def freeze(self):
+        """Freeze the generic collection and make it immutable going forward."""
+        self.frozen = True
+
+    @classmethod
+    def default(cls) -> 'GenericCollection':
```
Comment (Mikaayenson) [inline detection_rules/generic_loader.py:165]: Will there be unit tests that load this Generic Collection? Just thinking about how a user will test this prior to trying to push to kibana and loading everything.

Comment (brokensound77) [reply detection_rules/generic_loader.py:165]: Not built in - but if they wanted to unit test similar to how we do rules, they would just need to load and test

Comment (Mikaayenson) [reply detection_rules/generic_loader.py:165]: you mean use the loader to "test" in terms of schema validation, but no unit testing explicitly for field validation?

**Comment by Mikaayenson** [comment]
Do we need all the same metadata like rule information if a user wants to directly add schemas to the toml file?

**Pull request merged**